### PR TITLE
document Kubernetes Ingress Controller integration with Kong Mesh

### PR DESCRIPTION
**Pull request by ruromero**
Signed-off-by: ruromero <rromerom@redhat.com>

### CODE REVIEW

**Review by mpaulgreen — COMMENTED** [review]
@ruromero, I completely got lost in this PR. Basically we want to do everything in kong-gateway-operator.md and replace the step for sample service with front end service in kuma-demo ns documented in 
## Defining a Service and a Route. We can discuss this in our call tomorrow.

**Comment by mpaulgreen** [comment]
@ruromero If I recall correctly, for gateway we had a similar issue. We had to run the helm deployment with

--set postgresql.securityContext.runAsUser=1000710000 \
--set postgresql.securityContext.fsGroup= \

where 1000710000 was found from describing the ```kong``` namespace

**Comment by mpaulgreen** [comment]
@ruromero shared from Kong

helm install kong kong/kong -n kong \
--set ingressController.installCRDs=false \
--set ingressController.image.tag=2.3.1 \
--set podAnnotations."kuma\.io/mesh"=default \
--set podAnnotations."kuma\.io/gateway"=enabled

File: kong-kic.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+date: Thu, 28 Apr 2022 15:19:01 GMT
+server: envoy
+set-cookie: 157c7d417676c54695fa6cf886b2feeb=fde5bcb91b17e49f15ea3646b20eb9b6; path=/; HttpOnly
+via: kong/2.8.1
+x-kong-proxy-latency: 0
+x-kong-upstream-latency: 244
+
+upstream connect error or disconnect/reset before headers. reset reason: connection termination
+
+```
```
[the remark this answers is not in the document]
Comment (mpaulgreen) [reply kong-kic.md:69]: I kept it running for more than hour, it kept on working.

[the remark this answers is not in the document]
Comment (mpaulgreen) [reply kong-kic.md:69]: After I applied the mTLS I could still reach the frontend

[the remark this answers is not in the document]
Comment (mpaulgreen) [reply kong-kic.md:69]: I just tried. to open thru browser, nothing comes up.

**Review by mpaulgreen — COMMENTED** [review]
@ruromero I can reach the service by tweaking the steps you mentioned. I certainly get a 200OK  response from the frontend.
```
HTTP/1.1 200 OK
cache-control: max-age=3600
cache-control: private
content-length: 862
content-type: text/html; charset=UTF-8
date: Thu, 28 Apr 2022 23:10:06 GMT
etag: W/"287318801-862-2020-08-16T00:52:19.000Z"
last-modified: Sun, 16 Aug 2020 00:52:19 GMT
server: envoy
set-cookie: 157c7d417676c54695fa6cf886b2feeb=fc3d38df6fb99af81d8c6c1110fade31; path=/; HttpOnly
via: kong/2.8.1.0-enterprise-edition
x-envoy-upstream-service-time: 3
x-kong-proxy-latency: 0
x-kong-upstream-latency: 5

<!DOCTYPE html><html lang=en><head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=/favicon.png><title>Kuma Marketplace</title><link href=/css/app.92b097f8.css rel=preload as=style><link href=/css/chunk-vendors.96abecf8.css rel=preload as=style><link href=/js/app.5b2c045d.js rel=preload as=script><link href=/js/chunk-vendors.1eee84b8.js rel=preload as=script><link href=/css/chunk-vendors.96abecf8.css rel=stylesheet><link href=/css/app.92b097f8.css rel=stylesheet></head><body><noscript><strong>We're sorry but Kuma Marketplace doesn't work properly without JavaScript enabled. Please enable it to continue.</strong></noscript><div id=app></div><script src=/js/chunk-vendors.1eee84b8.js></script><script src=/js/app.5b2c045d.js></script></body></html>```. But the app cannot be seen on the browser

File: kong-kic.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpaulgreen) [reply kong-kic.md:69]: Can you implement the CP and DP as advised by Kong in Slack first. You will not side car injection for CP. You can also readthru demo.md. Once you reach this point, you can apply logging, traffic permission, tracing etc just to ensure everything works.

File: kong-kic.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+kubectl delete svc kong-kong-proxy
+kubectl apply -f kic/expose-kong-proxy.yaml
+```
+
+Allow communication between kong Ingress Controller and the mesh
+
+```
+oc apply -f kic/traffic_permission.yaml
+```
+
```
[the remark this answers is not in the document]
Comment (ruromero) [reply kong-kic.md:41]: this is required if when installing the mesh we deleted the `allow-all-default` trafficPermission

**Review by mpaulgreen — COMMENTED** [review]
@ruromero Do you want to update the instructions to include the GW CP and DP with DB or will you be creating a new instruction?

File: openshift-registry/README.md [hunk 3]
```diff
@@ -75,7 +75,7 @@ help you automating the process.
 From the [OpenShift documentation](https://docs.openshift.com/container-platform/4.10/openshift_images/managing_images/using-image-pull-secrets.html#images-allow-pods-to-reference-images-across-projects_using-image-pull-secrets)
 
 ```bash
-for i in system metrics logging tracing
+for i in system metrics logging tracing kic
 do
```
Comment (mpaulgreen) [inline openshift-registry/README.md:79]: @ruromero Why do we need to add kong-mesh-kic. I dont think there is a namespace for that.

File: kong-kic.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+```
+127.0.0.1 - - [29/Apr/2022:08:42:59 +0000] "GET /status HTTP/2.0" 500 42 "-" "Go-http-client/2.0"
+2022/04/29 08:43:02 [error] 2073#0: *135 [lua] api_helpers.lua:526: handle_error(): /usr/local/share/lua/5.1/lapis/application.lua:424: /usr/local/share/lua/5.1/kong/api/routes/health.lua:52: http2 requests not supported yet
+stack traceback:
+       [C]: in function 'capture'
+       /usr/local/share/lua/5.1/kong/api/routes/health.lua:52: in function 'fn'
+       /usr/local/share/lua/5.1/kong/api/api_helpers.lua:293: in function 'fn'
+        /usr/local/share/lua/5.1/kong/api/api_helpers.lua:293: in function </usr/local/share/lua/5.1/kong/api/api_helpers.lua:276>
+```
```
Comment (mpaulgreen) [inline kong-kic.md:77]: @ruromero I checked thru the logs for kong deployment. Did not see any errors related to it. Is it something specific to your environment?

File: openshift-registry/README.md (hunk 3 above)
Comment (ruromero) [reply openshift-registry/README.md:79]: this loop is not related to namespaces but to file names. I included a kong-kic.properties file containing the images that are used for Kubernetes Ingress Controller installation

**Review by mpaulgreen — COMMENTED** [review]
@ruromero I am keeping this PR open at this point for Claudio to get back to us on why the TrafficPermissions are not working. I did a multi-zone setup and it worked over there but not for gateway.

File: openshift-registry/README.md (hunk 3 above)
Comment (mpaulgreen) [reply openshift-registry/README.md:79]: But inside the loop you are  allowing pods to reference images across projects. Lets discuss this when you are back.

File: kong-kic.md (hunk 4 above)
Comment (mpaulgreen) [reply kong-kic.md:77]: I am using OCP 4.9.10

**Pull request merged**